---
title: "blockRAR: An R package for Simulation of Block Design for Response-Adaptive Randomization"
author: "Thevaa Chandereng, Rick Chappell"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{blockRAR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  \usepackage[utf8]{inputenc}
---
  
  
```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(
  echo = FALSE,
  collapse = TRUE,
  comment = "#>"
)
set.seed(43232)
```


# Introduction

Response-Adaptive Randomization (RAR) design utilizes accrual information to tilt the randomization ratio to the better performing treatment group.
Patients enrolled in these trials are not only treated to obtain effectiveness of a treatment but also treated to the best way possible. 
Altering the randomization ratio drastically can potentially affect the bias in a trial especially if one treatment is very superior to the other treatment. The major drawback to response-adaptive randomization design is that the trial needs to be short to be able to obtain the outcome of the trial for future randomization.
There are various methods developed to address some of this flaws besides developing a proper design for response-adaptive randomization.

Time drift can lead to bias by confounding the treatment effect with time effects.
Time trends are nearly universally ignored among RAR proponents.
Besides skewed randomization ratio, time trend can be a confounding factor as illustrated in the example below. 

```{r, echo = FALSE, fig=TRUE, fig.width = 6, fig.height = 6}
library(ggplot2)
set.seed(20999)


dat <- data.frame(
prob_response <- 0.5 + cumsum(c(0, runif(49, 0, 0.01))) + c(0, rnorm(49, 0, 0.01)),
time <- 0:49
)
dat$rand.ratio <- dat$prob_response


p <- ggplot(dat, aes(x = time))
p <- p + geom_line(aes(y = prob_response), color = "blue")
p <- p + scale_y_continuous(sec.axis = sec_axis(~., name = "Randomization Fraction for Treatment"))
p <- p + scale_colour_manual(values = c("blue"))
p <- p + labs(y = "Probability of Response (Clinical)",
              x = "Time (months)")
p <- p + theme(axis.text=element_text(size=12),
               axis.title=element_text(size=16,face="bold"))
p
```

Example of time trend issue in response-adaptive randomization (RAR) design:

 - The disease itself can change, sometimes radically (e.g., AIDS in the early 1990s).
- Our definition of the disease can change due to new scientific discoveries or diagnostic methods (e.g., stage migration in nasopharyngeal carcinoma due to the introduction of CAT scans to Hong Kong ~ 2005).
- Inclusion criteria can change, either formally (in which case we can stratify analysis on before vs. after the change) or informally due to “recruiting zeal” or other issues (in which case we can’t).
- Centers can change, such as when VAs enter the trial earlier or later than academic institutions.
- Patients within centers can change, especially but not only with chronic diseases, due to the phenomenon of “A queue of desperate patients lining up at the door”.
- In addition to these examples, an investigator who wants to game the system could cross his/her fingers that his favored treatment arm is ahead, then progressively enroll better prognosis patients over time

Therefore, we designed a stratified group-sequential method on altering the randomization ratio in a block/group level to address this issue instead of altering the randomization ratio by patient basis.
In each block/group, the randomization ratio is kept constant.
However, the ideal number of block/group is still vague.  


`blockRAR` is an R package for the simulation and analysis of adaptive Bayesian randomzied controlled trials under a range of trial designs and outcome types. Currently, it supports binomial data types, with time-to-event data and Gaussian data type to follow soon. The package allows the (optional) incorporation of historical data and permits interim analyses of early stopping for futility or success.

If you use `blockRAR` in published research, please cite: **TBA**.

If after reading through this vignette you have questions or problems using `blockRAR`, please post them to https://github.com/thevaachandereng/blockRAR/issues. This will notify the package maintainers and we can work to address the issue(s). Please note that comments and questions should **not** be emailed directly to the package authors.


# Running blockRAR

Prior to analyzing your data, the R package needs to be installed. The easiest way to install `blockRAR` is through CRAN:

```{r, eval = FALSE, echo=TRUE}
install.packages("blockRAR")
```

There are additional ways to download `blockRAR`. The first option is most useful for downloading a specific version of `blockRAR` (which can be found at https://github.com/thevaachandereng/bayesCT/releases):

```{r, eval = FALSE, echo=TRUE}
devtools::install_github("thevaachandereng/blockRAR@vx.xx.x")

# or 

devtools::install_version("blockRAR", version = "x.x.x", repos = "http://cran.us.r-project.org")
```

The second option is to download the most recent stable version through GitHub:

```{r, cache=FALSE, warning=FALSE, comment=FALSE, eval = FALSE, echo=TRUE, results="hide"}
devtools::install_github("thevaachandereng/blockRAR")
```

After successful installation, the package must be loaded into the working space:

```{r lib, results="asis", eval=TRUE, echo=TRUE}
library(blockRAR)
```

# Required input


## Frequentist Approach

**Data**:  Input to LPWC is a matrix or data frame.
We assume here that it is gene expression data, but other data types are suitable as well.
The expression matrix should be n -by- p where n is the number of genes (in rows) and p is the number of timepoints.
If **data** is a data frame, the gene names can be specified using the `row.names()`.

The object **simdata** is a simulated dataset for 200 genes with 8 timepoints that follow the biological [impulse model](https://doi.org/10.1093/bioinformatics/btw665).
This is stored as a data frame in the package with row names representing gene labels.




**Timepoints:** The object **timepoints** should be a vector of timepoints that specify when the data were collected.
This should match the column length in **Data**.
This can be in any form (seconds, minutes, hours, or even days), but make sure the units are uniform throughout.



**Data** and **Timepoints** are the main requirement to run LPWC, however, there are other optional parameters that affect the results.
Our manuscript referenced above describes these parameters in more detail.

**Max.lag:** The object **max.lag** should be an integer value that controls the maximum lag.
This max.lag has to be less than the floor of length(timepoints) / 4.
That is because longer lags lead to comparisons between two expression profiles that use less than half of the data points, which can lead to spurious correlations.

**C:** The object **C** should be a numeric.
This controls the width of the Gaussian kernel function used to penalize the lag and the weights in weighed correlation vector.
The parameter **C** can be set automatically by setting **C** to the default NULL value and setting the **penalty** argument described below.
If **C** is set to a specific value, the **penalty** is ignored.

**penalty:** The object **penalty** only allows two values: "low" and "high".
It is used to automatically set the value of **C**.
High imposes a higher penalty on lags thus favoring grouping genes without introducing lags.
The lower penalty leads to more lagged genes in the clusters.
The default is set to the "high" penalty.
This argument is ignored if **C** is set to a specific value.

**iter:** The object **iter** should be a numeric.
This controls the number of values of C that are tested when automatically setting the C parameter.
The default is 10.
Increasing iter increases computational time.
This is only relevant for the low penalty because the high penalty uses a single C instead of searching over multiple C.

